src/mrinufft/operators/off_resonance.py

```diff
@@ -0,0 +1,325 @@
+    C : np.ndarray or GPUarray
+        Off-resonance phase map at each time segment center of shape
+        ``(n_time_segments, *fieldmap.shape)``.
+    """
+    # get backend and device
+    xp = get_array_module(fieldmap)
+
+    if xp.__name__ == "torch":
+        is_torch = True
+        if fieldmap.device.type == "cpu":
+            xp = np
+            fieldmap = fieldmap.numpy(force=True)
+        else:
+            assert CUPY_AVAILABLE, "GPU computation requires Cupy!"
+            xp = cp
+            fieldmap = cp.from_dlpack(fieldmap)
+    else:
+        is_torch = False
+
+    # move t to backend
+    fieldmap = xp.asarray(fieldmap, dtype=xp.complex64)
+    t = xp.asarray(t, dtype=xp.float32).ravel()
+
+    # default
+    if isinstance(n_bins, (list, tuple)) is False:
+        n_bins = (n_bins, 10)
+
+    # transform to list
+    n_bins = list(n_bins)
+
+    # get field map
+    if xp.isreal(fieldmap).all().item():
+        r2star = None
+        b0 = fieldmap
+        fieldmap = 0.0 + 1j * b0
+    else:
+        r2star = fieldmap.real
+        b0 = fieldmap.imag
+
+    # default mask
+    if mask is None:
+        mask = xp.ones_like(fieldmap, dtype=bool)
+
+    # Hz to radians / s
+    fieldmap = 2 * math.pi * fieldmap
+
+    # create histograms
+    if r2star is not None:
+        z = fieldmap[mask].ravel()
+        z = xp.stack((z.imag, z.real), axis=1)
+        hk, ze = xp.histogramdd(z, bins=n_bins)
+        ze = list(ze)
+
+        # get bin centers
+        zc = [e[1:] - (e[1] - e[0]) / 2 for e in ze]
+
+        # complexify
+        zk = _outer_sum(1j * zc[0], zc[1])  # [K1 K2]
+        zk = zk.T
+        hk = hk.T
+    else:
+        z = fieldmap[mask].ravel()
+        hk, ze = xp.histogram(z.imag, bins=n_bins[0])
+
+        # get bin centers
+        zc = ze[1:] - (ze[1] - ze[0]) / 2
+
+        # complexify
+        zk = 0 + 1j * zc  # [K 1]
+
+    # flatten histogram values and centers
+    hk = hk.ravel()
+    zk = zk.ravel()
+
+    # generate time for each segment
+    tl = xp.linspace(
+        t.min(), t.max(), n_time_segments, dtype=xp.float32
+    )  # time seg centers in [s]
+
+    # prepare for basis calculation
+    ch = xp.exp(-tl[:, None, ...] @ zk[None, ...])
+    w = xp.diag(hk**0.5)
+    p = xp.linalg.pinv(w @ _transpose(ch)) @ w
+
+    # actual temporal basis calculation
+    B = p @ xp.exp(-zk[:, None, ...] * t[None, ...])
+    B = B.astype(xp.complex64)
+
+    # get spatial coeffs
+    C = xp.exp(-tl * fieldmap[..., None])
+    C = C[None, ...].swapaxes(0, -1)[
+        ..., 0
+    ]  # (..., n_time_segments) -> (n_time_segments, ...)
+
+    # clean-up of spatial coeffs
+    C = xp.nan_to_num(C, nan=0.0, posinf=0.0, neginf=0.0)
+
+    # back to torch if required
+    if is_torch:
+        if xp.__name__ == "cupy":
+            B = torch.from_dlpack(B)
+            C = torch.from_dlpack(C)
+        else:
+            B = torch.from_numpy(B)
+            C = torch.from_numpy(C)
+
+    return B, C
```

Comment:
I think we have `with_numpy_cupy` function which internally converts everything to numpy or cupy. Feel free to use it to prevent redundancy:

https://github.com/mind-inria/mri-nufft/blob/aad8c0e77f6c059d139e2ba37ce7f78e0914a787/src/mrinufft/operators/base.py#L178-L216

Reply:
I tried before, but the expected signature for "with_numpy_cupy" (data, output=None, *args, **kwargs) seems to not work well with get_interpolators_from_fieldmap (while I guess it is generally ok for nufft routines).

Reply:
A solution is to modify `with_numpy_cupy`: remove the output named arg and instead check if it is in the `**kwargs`. You would get a generic enough function to suit your  (and potentially other's) needs.

Reply:
Ok, this is a bit more complicated than I expected. The decorator is mainly designed for class methods, and expect that "self" is passed. For standalone functions, "data" will be the second argument of the wrapper. I could:

1) move `get_interpolators_from_fieldmap` inside MRIFourierCorrected.
2) modify the decorator to inspect the signature and see if "self" or "cls" are detected as first argument, and branch between the class and standalone versions
3) add a "classmethod = False" argument to the decorator and switch between the two brances according to this flag. This would be more robust than 2) as both self/cls are conventions and might not be always respected - setting the default flag as False should make it backward compatible with the existing code.

1) is the cleanest, but maybe it would be nice to be able to decorate standalone functions? what do you think?

Reply:
You raise a valid point. I would say 2) would be the long term solution (using `inspect.ismethod()` , and refactoring all the `with_` decorators we have build). This is not the purpose of this PR, so let's just focus on the B0 stuff for now, and keep your working code as is.

Reply:
Ok, I have done it in a different branch, I'll open a PR after this is merged